ppl/zqd/recruiter/recruiter.go

```diff
@@ -0,0 +1,191 @@
+package recruiter
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+// WorkerPool is the internal state of the recruiter system
+// The pools are exported only for use in unit tests
+type WorkerPool struct {
+	mu           sync.Mutex                // one lock for all three maps
+	freePool     map[string]WorkerDetail   // Map of all free workers
+	nodePool     map[string][]WorkerDetail // Map of nodes of slices of free workers
+	reservedPool map[string]WorkerDetail   // Map of busy workers
+	SkipSpread   bool                      // option to test algorithm performance
+}
+
+type WorkerDetail struct {
+	Addr     string
+	NodeName string
+}
+
+func NewWorkerPool() *WorkerPool {
+	rand.Seed(time.Now().UnixNano()) // for shuffling nodePool keys
+	return &WorkerPool{
+		freePool:     make(map[string]WorkerDetail),
+		nodePool:     make(map[string][]WorkerDetail),
+		reservedPool: make(map[string]WorkerDetail),
+	}
+}
+
+func (pool *WorkerPool) Register(addr string, nodename string) error {
+	if _, _, err := net.SplitHostPort(addr); err != nil {
+		return fmt.Errorf("Invalid address for Register: %w", err)
+	}
+	if nodename == "" {
+		return fmt.Errorf("Node name required for Register")
+	}
+	wd := WorkerDetail{Addr: addr, NodeName: nodename}
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	_, prs := pool.reservedPool[addr]
+	if prs {
```

Comment:
Nits: We usually use the name ok for the second value produced by a map index expression. And we usually do this on one line.
```suggestion
	if _, ok := pool.reservedPool[addr]; ok {
```